example/src/main/java/org/xmtp/android/example/pushnotifications/PushNotificationsService.kt

```diff
@@ -0,0 +1,101 @@
+import org.xmtp.android.example.extension.truncatedAddress
+import org.xmtp.android.library.messages.EnvelopeBuilder
+import java.util.Date
+
+class PushNotificationsService : FirebaseMessagingService() {
+
+    companion object {
+        private const val TAG = "PushNotificationService"
+
+        internal const val CHANNEL_ID = "xmtp_message"
+    }
+
+    override fun onNewToken(token: String) {
+        super.onNewToken(token)
+        PushNotificationTokenManager.syncPushNotificationsToken(token)
+    }
+
+    override fun onMessageReceived(remoteMessage: RemoteMessage) {
+        super.onMessageReceived(remoteMessage)
+        Log.d(TAG, "On message received.")
+
+        val keysData = loadKeys()
+        if (keysData == null) {
+            Log.e(TAG, "Attempting to send push to a logged out user.")
+            return
+        }
+
+        val encryptedMessage = remoteMessage.data["encryptedMessage"]
+        val topic = remoteMessage.data["topic"]
+        val encryptedMessageData = Base64.decode(encryptedMessage, Base64.NO_WRAP)
+        if (encryptedMessage == null || topic == null || encryptedMessageData == null) {
+            Log.e(TAG, "Did not get correct message data from push")
+            return
+        }
+
+        GlobalScope.launch(Dispatchers.Main) {
+            ClientManager.createClient(keysData)
+        }
+        val conversation = ClientManager.client.fetchConversation(topic)
+        if (conversation == null) {
+            Log.e(TAG, "No keys or conversation persisted")
+            return
+        }
+        val envelope = EnvelopeBuilder.buildFromString(topic, Date(), encryptedMessageData)
+        val peerAddress = conversation.peerAddress
+        val decodedMessage = conversation.decode(envelope)
+
+        val body = decodedMessage.body
+        val title = peerAddress.truncatedAddress()
+
+        val pendingIntent = PendingIntent.getActivity(
+            this,
+            0,
+            ConversationDetailActivity.intent(
+                this,
+                topic = topic,
+                peerAddress = peerAddress
+            ),
+            (PendingIntent.FLAG_IMMUTABLE or PendingIntent.FLAG_UPDATE_CURRENT)
+        )
+
+        val builder = NotificationCompat.Builder(this, CHANNEL_ID)
+            .setSmallIcon(R.drawable.ic_xmtp_white)
+            .setContentTitle(title)
+            .setContentText(body)
+            .setAutoCancel(true)
+            .setColor(ContextCompat.getColor(this, R.color.black))
+            .setPriority(NotificationCompat.PRIORITY_DEFAULT)
+            .setStyle(NotificationCompat.BigTextStyle().bigText(body))
+            .setContentIntent(pendingIntent)
+
+        // Use the URL as the ID for now until one is passed back from the server.
+        NotificationManagerCompat.from(this).apply {
+            notify(topic.hashCode(), builder.build())
+        }
+    }
+
+    private fun loadKeys(): String? {
```

Comment:
TINY suggestion: it might be nice to share this method with `MainActivity`'s loadKeys() but def not blocking and up to you!